mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java

```diff
@@ -19,21 +19,408 @@
+    }
+
+    @Override
+    public void write(Writable row) {
+      Preconditions.checkArgument(row instanceof IcebergWritable);
+
+      // TODO partition handling
+      appender.add(((IcebergWritable) row).record());
+    }
+
+    @Override
+    public void write(NullWritable key, IcebergWritable value) {
+      write(value);
+    }
+
+    @Override
+    public void close(boolean abort) throws IOException {
+      appender.close();
+      if (!abort) {
+        fileData.put(taskAttemptId, new ClosedFileData(location, fileFormat, appender.length(), appender.metrics()));
+      }
+    }
+
+    @Override
+    public void close(org.apache.hadoop.mapreduce.TaskAttemptContext context) throws IOException {
+      close(false);
+    }
+
+    @Override
+    public void close(Reporter reporter) throws IOException {
+      close(false);
+    }
+  }
+
+  /**
+   * A dummy committer - not related to the Hive transactions.
+   */
+  public static final class IcebergOutputCommitter extends OutputCommitter {
+
+    @Override
+    public void setupJob(JobContext jobContext) {
+      // do nothing.
+    }
+
+    @Override
+    public void setupTask(TaskAttemptContext taskAttemptContext) {
+      // do nothing.
+    }
+
+    @Override
+    public boolean needsTaskCommit(TaskAttemptContext taskAttemptContext) {
+      return true;
+    }
+
+    @Override
+    public void commitTask(TaskAttemptContext taskAttemptContext) throws IOException {
+      ClosedFileData closedFileData = fileData.remove(taskAttemptContext.getTaskAttemptID().toString());
+      createCommittedFileFor(new HadoopFileIO(taskAttemptContext.getJobConf()), closedFileData);
+    }
+
+    @Override
+    public void abortTask(TaskAttemptContext taskAttemptContext) {
+      fileData.remove(taskAttemptContext.getTaskAttemptID().toString());
+    }
+
+    @Override
+    public void commitJob(JobContext jobContext) throws IOException {
+      Configuration conf = jobContext.getJobConf();
+      Path queryResultPath = new Path(generateQueryLocation(conf));
+      Table table = Catalogs.loadTable(conf);
+
+      ExecutorService executor = null;
+      try {
+        // Creating executor service for parallel handling of file reads and deletes
+        executor = Executors.newFixedThreadPool(
+            conf.getInt(InputFormatConfig.WRITE_THREAD_POOL_SIZE, InputFormatConfig.WRITE_THREAD_POOL_SIZE_DEFAULT),
+            new ThreadFactoryBuilder()
+                .setDaemon(false)
+                .setPriority(Thread.NORM_PRIORITY)
+                .setNameFormat("iceberg-commit-pool-%d")
+                .build());
+
+        Set<String> taskTmpFiles = new HashSet<>();
+        Set<Future<DataFile>> dataFiles = new HashSet<>();
+        FileSystem fs = Util.getFs(queryResultPath, conf);
+
+        // Listing the task result directory and reading .committed files
+        RemoteIterator<LocatedFileStatus> taskFileStatuses = fs.listFiles(queryResultPath, false);
```

Comment:
Iceberg writes can't depend on listing files, or else the commit may produce incorrect results because files are left out of the listing.

I think we can work around this by using well-known file paths. If we know that there will be 4 write tasks, then we can look for 4 files located at `(jobTempPath)/(jobId)/task-(num)`: `task-0`, `task-1`, `task-2`, and `task-3`. If any file isn't found then the commit fails. This requires knowing how many tasks were used, but I think that it should be possible to find out how many write tasks were run.

Reply:
I did the proposed changes.
I have the following reservations:
- The assumption is that the taskIds are generated in sequential order starting from 0. - I am not sure about this. Any idea how can I confirm this?
- When a reducer ends up not writing any data then no recordWriter is created/closed, and no data is written to the disk. Will try to find a way to mark this job as a committed as well with an empty dataset.

Reply:
Have a solution for handling tasks not writing actual data. The question still remains if the taskIds are generated in a sequential order from 0 or not.

Reply:
I've only ever seen task IDs that are increasing from 0, but I couldn't point to the code where they are assigned.

Reply:
Let's close this now, and check again it if needed